server/controllers/apiController.js

```diff
@@ -4,60 +4,76 @@
-  var regex = new RegExp(/(good|great|awesome|fantastic|terrific|nice|cool|wonderful|dope|beautiful|amazing|gorgeous|breathtaking) view/, 'i');
+  var searchString = urlParser.parse(req.url).search; //include leading question mark
+  var regex = new RegExp(/(good|great|awesome|fantastic|terrific|nice|cool|wonderful|dope|beautiful|amazing|gorgeous|breathtaking) view/);
 
   request.get('https://maps.googleapis.com/maps/api/place/nearbysearch/json' + searchString + '&key=' + GOOGLE_PLACES_API_KEY)
-    .on('response', function(response) {
-      console.log(response.body.results);
-      var filteredBody = {};
-      filteredBody.places = [];
-      var places = response.body.results;
-      var counter = 0; //ensures server only sends back filteredBody if all places have been processed
-      for (var i = 0; i < places.length; i++) {
-        var place = places[i];
-        var placeid = place['place_id'];
+    .on('response', function(response) { //layer 1 on 'response'
+
+      var body = [];
 
-        request.get('https://maps.googleapis.com/maps/api/place/details/json?' + 'key=' + GOOGLE_PLACES_API_KEY + '&placeid=' + placeid)
-          .on('response', function(response) {
-            var placeDetails = response.body.result;
-            var reviews = placeDetails.reviews;
-            for (var j = 0; j < reviews.length; j++) {
-              var review = reviews[j];
-              if (review.text.match(regex)) { //TODO: improve regex matching
-                filteredBody.places.push({
-                  name: placeDetails.name,
-                  address: placeDetails['formatted_address']
-                });
-                break;
-              }
-            }
-            counter++;
-            if (counter === places.length) {
-              res.json(filteredBody); //also needs to pass page token back to client for further requests
-            } //TODO: move it inside callback above
-          })
-          .on('error', function(error) {
-            //TODO: handle error
-            counter++;
-            if (counter === places.length) {
-              res.json(filteredBody); //also needs to pass page token back to client for further requests
-            } 
-          })
+      response.on('data', function(chunk) { //layer 2 on 'data'
+        body.push(chunk);
+      }).on('end', function() { //layer 2 on 'end'
+        body = JSON.parse(Buffer.concat(body).toString());
+        var filteredBody = {};
+        filteredBody.places = [];
+        var places = body.results;
+        var counter = 0; //ensure server only sends back filteredBody if all places have been processed
+        for (var i = 0; i < places.length; i++) {
+          var place = places[i];
+          var placeid = place['place_id'];
 
-    })
-    .on('error', function(error) {
+          request.get('https://maps.googleapis.com/maps/api/place/details/json?' + 'key=' + GOOGLE_PLACES_API_KEY + '&placeid=' + placeid)
```

Comment:
I think this would benefit from more whitespace and comments summarizing each layer of the callbacks.